Add error context

Harfusha opened the pull request:
https://github.com/cakephp/cakephp/issues/17975

Harfusha committed:
Add error context

Harfusha committed:
Revert original trace, fix annotation

Harfusha committed:
CsFix, fix annotation

dereuromark commented on src/Error/ErrorLogger.php:
```diff
@@ -85,6 +84,31 @@ public function logError(PhpError $error, ?ServerRequestInterface $request = nul
+     * @param bool $includeTrace Whether or not to include a stack trace.
+     * @return string Error message
+     */
+    private function getErrorMessage(PhpError $error, bool $includeTrace = false): string
```
It might be better to move this to 5.next and make this protected instead.
But for me either way works for now.

Harfusha replied:
Change from private to protected was planned, but forgotten :D

dereuromark replied:
But then it must go into 5.next with a note in the migration guide IMO.

Harfusha committed:
Change getErrorMessage from private to protected

ADmad commented:
Can you give an e.g. of error message without file and line number?

I already see them in the error messages in the log file, for e.g.:
```
<date-here> warning: Could not render cell - Class `App\View\Cell\NavCell` does not have a `foo` method. [<redacted>/vendor/cakephp/cakephp/src/View/Cell.php, line 178]
```

Harfusha commented:
Old:
```
2024-10-17 05:33:43 warning: A warning caught by ErrorTrap
Request URL: /cart/calculator
Referer URL: ___
Client IP: ___
Trace:
/var/www/adam/dev/src/Controller/AppController.php /var/www/adam/dev/src/Controller/AppController.php, line 141
...
```

New:
```
2024-10-17 05:33:50 warning: A warning caught by ErrorTrap in /var/www/adam/dev/src/Controller/AppController.php on line 141
Stack Trace:
- /var/www/adam/dev/src/Controller/AppController.php:141
...
- [main]:

Request URL: /
Client IP: ___
```

Triggered by:
`trigger_error("A warning caught by ErrorTrap", E_USER_WARNING);`

Harfusha commented:
This was inconsistent between the log messages

ADmad commented:
So you are directly using the PHP function instead of the Cake function `triggerWarning()` which already includes the file and line number. Your change will duplicate that info for calls to `triggerWarning()` which isn't acceptable either.

Harfusha committed:
CsFix

Harfusha commented:
Not me, but libs i use. 

+ This will make it same as normal warnings:
```
2024-10-17 06:35:00 error: [TypeError] in_array(): Argument #2 ($haystack) must be of type array, string given in /var/www/adam/dev/src/Controller/AppController.php on line 141
Stack Trace:
- /var/www/adam/dev/src/Controller/AppController.php:141
...
- [main]:

Request URL: /
Client IP: ___
```

Will add fix for the duplicity

Harfusha commented:
Oh i see the problem. Each trigger_error used in cakephp has the file and line added where they are called.

I would like to make some function that will format it same way (When it is called or when it is outside cakephp, in ErrorTrap), so it can be parsed programatically. Now each type of error has own format.

Any ideas?

markstory commented:
> I would like to make some function that will format it same way (When it is called or when it is outside cakephp, in ErrorTrap), so it can be parsed programatically. Now each type of error has own format.

The file/line could be removed from `triggerWarning()` and included in the output generated by `ErrorTrap`. The logic in `triggerWarning()` predates `ErrorTrap` by quite a bit of time, and we should be able to include file/line in rendered output more consistently now.

dereuromark commented:
Sounds like this should go into 5.next then, right? With such changes.

ADmad commented:
This doesn't change the public API and making the log output consistent is more of a bigfix than a new feature.

Harfusha commented:
I will try to make it next week :)

dereuromark commented:
@Harfusha The next patch release is imminent in case you want to fit it into that one still.

Harfusha commented:
I dont have any time to do this right now. I will do it when i will have free time to do this

dereuromark commented:
Ping @Harfusha

Harfusha commented:
I'm very sorry but i don't have any free time anytime soon.

dereuromark commented:
@markstory what's needed here to continue? Maybe the core team could finish this one up.

markstory committed:
Remove redundant error logging context

with error logging context being added by ErrorTrap, we don't need to do
it within `triggerError` which is now a simple alias.

markstory committed:
Merge branch '5.x' into context-in-error

markstory commented:
@dereuromark Updated with what I had suggested. It does change behavior a bit as `triggerWarning()` becomes a simple wrapper which changes its output outside of applications that also use `ErrorTrap`.

dereuromark commented:
Do we want to move forward with this?

markstory committed:
Fix test for functionsglobaltest as well

markstory committed:
Fix phpcs
